extensions/Mist/KeyHistory.js

```diff
@@ -0,0 +1,151 @@
+// Then you can obtain one at https://mozilla.org/MPL/2.0/
+
+(function (Scratch) {
+  "use strict";
+
+  class KeyHistoryExtension {
+    constructor() {
+      this.keyHistory = [];
+      this.pasted = false;
+      this.max_key_history = 100; // Adjust the maximum number of keys to keep in history
+      this.keybinds = ["Ctrl", "Shift", "Alt"];
+      this.pause = false;
+    }
+
+    getInfo() {
+      return {
+        id: "KeyHistoryExtension",
+        name: "Key History",
+        color1: "#36644E",
+        blocks: [
+          {
+            opcode: "getRecentKeys",
+            blockType: Scratch.BlockType.REPORTER,
+            text: "Get recent keys",
+          },
+          {
+            opcode: "deleteFirstKey",
+            blockType: Scratch.BlockType.COMMAND,
+            text: "Delete the first key from history",
+          },
+          {
+            opcode: "deleteAllKeys",
+            blockType: Scratch.BlockType.COMMAND,
+            text: "Delete all keys from history",
+          },
+          {
+            opcode: "AddKey",
+            blockType: Scratch.BlockType.COMMAND,
+            text: "Add [KEY] to key history",
+            arguments: {
+              KEY: {
+                type: Scratch.ArgumentType.STRING,
+                defaultValue: " ",
+              },
+            },
+          },
+          {
+            opcode: "enableKeyHistory",
+            blockType: Scratch.BlockType.COMMAND,
+            text: "Enable key history",
+          },
+          {
+            opcode: "disableKeyHistory",
+            blockType: Scratch.BlockType.COMMAND,
+            text: "Disable key history",
+          },
+        ],
+      };
+    }
+
+    getRecentKeys() {
+      return JSON.stringify(this.keyHistory);
+    }
+
+    deleteFirstKey() {
+      if (this.keyHistory.length > 0) {
+        this.keyHistory.shift();
+      }
+    }
+
+    deleteAllKeys() {
+      this.keyHistory = [];
+    }
+
+    AddKey({ KEY }) {
+      this.addKeyToHistory(KEY);
+    }
+
+    onKeyDown(event) {
+      // Check if Command (Cmd) or Control (Ctrl) keys are pressed
+      if (event.metaKey || event.ctrlKey) {
+        return; // Skip adding keys when Cmd or Ctrl are pressed
+      }
+
+      // Check if the pressed key is part of a keybind
+      if (this.isKeybind(event.key)) {
+        return; // Skip adding keybind keys to history
+      }
+
+      // Add the pressed key to the history
+      if (event.key && !this.pause) {
+        const key = event.key;
+        this.addKeyToHistory(key);
+      }
+    }
+
+    onPaste(event) {
+      const pastedText = event.clipboardData.getData("text/plain");
+      if (pastedText.trim() !== "") {
+        if (this.pasted == false) {
+          this.pasted = true;
+          this.addKeyToHistory(pastedText);
+        }
+        setTimeout(() => {
+          this.pasted = false;
+        }, 10);
+      }
+    }
+
+    isKeybind(key) {
+      return this.keybinds.includes(key);
```

Comment:
made it so that inputs are converted

This is an example v
![Screenshot 2024-07-30 at 22 26 35](https://github.com/user-attachments/assets/fed68787-eba9-4095-a111-3c127c37efe8)